cmd/kubeadm/app/cmd/init.go

```diff
@@ -163,6 +163,24 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
 		if len(initRunner.Options.SkipPhases) == 0 {
 			initRunner.Options.SkipPhases = data.cfg.SkipPhases
 		}
+		// dump DNS and Proxy status by skip-phases
+		if isStringInSlice("addon/coredns", initRunner.Options.SkipPhases) {
+			data.cfg.DNS.Disabled = true
+		}
+		if isStringInSlice("addon/kube-proxy", initRunner.Options.SkipPhases) {
+			data.cfg.Proxy.Disabled = true
+		}
+		if isStringInSlice("addon", initRunner.Options.SkipPhases) {
+			data.cfg.DNS.Disabled = true
+			data.cfg.Proxy.Disabled = true
+		}
+		// If DNS or Proxy is disabled, skip the corresponding phase.
+		if data.cfg.DNS.Disabled {
+			initRunner.Options.SkipPhases = append(initRunner.Options.SkipPhases, "addon/coredns")
+		}
+		if data.cfg.Proxy.Disabled {
+			initRunner.Options.SkipPhases = append(initRunner.Options.SkipPhases, "addon/kube-proxy")
+		}
```

Comment:
all these nwe changes in SetDataInitializer should be moved to a new function - e.g.
```
manageSkippedAddons(cfg *ClusterConfiguration, skipPhases []string) 
```

and it should be unit tested.

Reply:
Done. Moved it to a new function `manageSkippedAddons` and added some unittest.